homeassistant/components/fritz/config_flow.py

```diff
@@ -0,0 +1,244 @@
+            return self._show_setup_form_confirm()
+
+        errors = {}
+
+        self._username = user_input.get(CONF_USERNAME)
+        self._password = user_input.get(CONF_PASSWORD)
+
+        error = await self.fritz_tools_init()
+
+        if error:
+            errors["base"] = error
+            return self._show_setup_form_confirm(errors)
+
+        return self._async_create_entry()
+
+    def _show_setup_form_init(self, errors=None):
+        """Show the setup form to the user."""
+        return self.async_show_form(
+            step_id="start_config",
+            data_schema=vol.Schema(
+                {
+                    vol.Optional(CONF_HOST, default=DEFAULT_HOST): str,
+                    vol.Optional(CONF_PORT, default=DEFAULT_PORT): vol.Coerce(int),
+                    vol.Required(CONF_USERNAME): str,
+                    vol.Required(CONF_PASSWORD): str,
+                }
+            ),
+            errors=errors or {},
+        )
+
+    def _show_setup_form_confirm(self, errors=None):
+        """Show the setup form to the user."""
+        return self.async_show_form(
+            step_id="confirm",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(CONF_USERNAME): str,
+                    vol.Required(CONF_PASSWORD): str,
+                }
+            ),
+            description_placeholders={"name": self._name},
+            errors=errors or {},
+        )
+
+    async def async_step_user(self, user_input=None):
+        """Handle a flow initiated by the user."""
+        return await self.async_step_start_config()
+
+    async def async_step_start_config(self, user_input=None):
+        """Handle a flow start config."""
+        if user_input is None:
+            return self._show_setup_form_init()
+
+        errors = {}
+
+        self._host = user_input.get(CONF_HOST, DEFAULT_HOST)
+        self._port = user_input.get(CONF_PORT, DEFAULT_PORT)
+        self._username = user_input.get(CONF_USERNAME)
+        self._password = user_input.get(CONF_PASSWORD)
+
+        error = await self.fritz_tools_init()
+
+        if not error:
+            self._name = self.fritz_tools.device_info["model"]
+
+            for entry in self._async_current_entries(include_ignore=False):
+                if entry.data[CONF_HOST] == self._host:
+                    error = "already_configured"
+
+        if error:
+            errors["base"] = error
+            return self._show_setup_form_init(errors)
+
+        return self._async_create_entry()
+
+    async def async_step_reauth(self, data):
+        """Handle flow upon an API authentication error."""
+        self._entry = self.hass.config_entries.async_get_entry(self.context["entry_id"])
+        self._host = data.data.get(CONF_HOST, DEFAULT_HOST)
+        self._port = data.data.get(CONF_PORT, DEFAULT_PORT)
+        self._username = data.data.get(CONF_USERNAME)
+        self._password = data.data.get(CONF_PASSWORD)
```

Comment:
The reauth is initialized with the entry data 
https://github.com/home-assistant/core/blob/dev/homeassistant/config_entries.py#L541

```suggestion
        self._host = data[CONF_HOST]
        self._port = data[CONF_PORT]
        self._username = data[CONF_USERNAME]
        self._password = data[CONF_PASSWORD]
```